### PR TITLE
Trainable Keypoint Detection Model

### PR DESCRIPTION
**Pull request by yycho0108**
After quite a few struggles, the keypoint detection model is finally able to train.
However, I have no idea (as of yet) if things are actually working = producing the output that we want.
Hopefully, once we incorporate a much more robust logging system, the situation would improve.

### Current State
Current output looks like the following:
```
$ python3 keypoint_regression.py --dataset CUBE --force_rebuild=False

/home/jamiecho/.cache/pypoetry/virtualenvs/top-ME4FQeLv-py3.6/lib/python3.6/site-packages/torch/nn/functional.py:1639: UserWarning: nn.functional.sigmoid is deprecated. Use torch.sigmoid instead.
  warnings.warn("nn.functional.sigmoid is deprecated. Use torch.sigmoid instead.")
tensor(0.7688, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.6957, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.5755, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.5282, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.5116, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.4887, device='cuda:0', grad_fn=<AddBackward0>)
```
### Updates
* Dataset Options from CLI
* Working dataset loading logic
* Working two-head keypoint loss
* Implement vertex displacement regression
* Implement heatmap distribution prediction
* Working `keypoint_regression.py`
* Apply `Schema` in a select few places that required updates

### TODOs:
* TQDM-style metrics logging
* Integrate TensorBoard logger into training
* Visualizations for the output bounding box predictions
* Support low-resolution predictions
* Consider incorporating keypoint regression in the style of [DSNTNN](https://github.com/anibali/dsntnn/tree/master/dsntnn)
* Collation function that either drops non-stackable args or pad inputs with `MAX_NUM_OBJECTS` inputs
* Finish global updates to use `Schema`
* Be consistent with our `keypoint` ordering and definitions (i.e. the index of the `CENTROID`, should it exist)
* Deal with `torch_xla` related issues
* Add documentation for all new functions and classes
* Apply regularization for network weights

### CODE REVIEW

File: scripts/keypoint_regression.py [hunk 1]
```diff
@@ -14,81 +17,132 @@
...
+            opts.sample_objectron, objectron=objectron_opts)
+        train_dataset = SampleObjectron(data_opts, transform)
+
+        objectron_opts = replace(
+            opts.sample_objectron.objectron, train=False)
+        data_opts = replace(
+            opts.sample_objectron, objectron=objectron_opts)
+        test_dataset = SampleObjectron(data_opts, transform)
+    else:
+        raise ValueError(F'Invalid dataset choice : {opts.dataset}')
```
Comment (ajy8456) [inline scripts/keypoint_regression.py:92]: I'm not familiar with dataclass yet, so could you please explain what happens with `replace()`?

File: src/top/data/colored_cube_dataset.py [hunk 2]
```diff
@@ -209,19 +213,32 @@ def __iter__(self):
...
+            if self.opts.unstack:
+                for i in range(self.opts.batch_size):
+                    out_i = {k: v[i] for k, v in out.items()}
+                    if self.xfm is not None:
+                        out_i = self.xfm(out_i)
+                    yield out_i
+            else:
+                if self.xfm is not None:
+                    out = self.xfm(out)
+                yield out
```
Comment (ajy8456) [inline src/top/data/colored_cube_dataset.py:241]: What exactly does `unstack` mean? Is it correct to understand that it is a process of yielding images one by one when batch size images are bundled and processed?

**Review by ajy8456 — APPROVED** [review]
It looks good except for two minor points related to my understanding.

File: scripts/keypoint_regression.py (hunk 1 above)
Comment (yycho0108) [reply scripts/keypoint_regression.py:92]: If there's a setting that's like `s={"a":1, "b":2}`, then `replace(s, a=3)` would make the resulting output `s={"a":3, "b":2}`. `replace` is required since generally we consider the `dataclass` instance itself to be immutable. This piece of code *is* quite confusing so I'll leave comments for this.

File: src/top/data/colored_cube_dataset.py (hunk 2 above)
Comment (yycho0108) [reply src/top/data/colored_cube_dataset.py:241]: Yep. The thing here is we're rendering a batch of images, yet `Objectron` yields images one by one, thus we just make the dataset also yield images one by one for compatibility.

File: scripts/keypoint_regression.py (hunk 1 above)
Comment (ajy8456) [reply scripts/keypoint_regression.py:92]: Oh, I got it, thank you.
